### PR TITLE
Implement new App Launcher Not Running Page/Functionality

### PR DESCRIPTION
**Pull request by jbouder**
## Reference Issues or PRs

N/A

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [x] Did you test the pull request locally?
- [x] Did you add new tests?

## Screenshots
Screenshot 2024-07-11 at 1 31 51 PM

## Any other comments?
- Updated to base page templates to simplify future legacy migrations
- Added new Not Running page and modal to React App
- Updated to redirect to home and open modal when app isn't started and user navigates to stopped app directly
- Updated app cards to display Not Running modal on click when not started

_Note: native browser-based session storage is being used to track redirect and open ux. This will be replaced with global state when app is running as full SPA_

### CODE REVIEW

File: jhub_apps/service/japps_routes.py [hunk 1]
```diff
@@ -23,7 +23,7 @@ async def handle_apps(request: Request):
     if not theme:
         theme = themes.DEFAULT_THEME
     return templates.TemplateResponse(
-        "japps_home.html",
+        "japps_custom.html",
```
Comment (jbouder) [inline jhub_apps/service/japps_routes.py:26]: Note, this page template is only used by pages which require a custom route through the japps service. This change was just to reduce confusion with naming between this template and the real base template (japps_page)

File: jhub_apps/templates/style.css [hunk 2]
```diff
@@ -383,6 +383,10 @@ a:focus {
     border: unset;
 }
 
+.card-dialog-body-wrapper {
+    padding: 0 24px 20px !important;
+}
+
```
Comment (jbouder) [inline jhub_apps/templates/style.css:389]: Fix spacing issue in modal dialogs, which are currently being overridden by JupyterLab styling

File: ui/src/components/context-menu/context-menu.tsx [hunk 3]
```diff
@@ -125,6 +125,7 @@ export const ContextMenu = ({
               }}
               onClick={(e) => {
                 if (!item.disabled && item.onClick) {
+                  e.stopPropagation();
```
Comment (jbouder) [inline ui/src/components/context-menu/context-menu.tsx:128]: This prevents the onclick event from bubbling up when the context menu should actually be handling the event

File: ui/src/pages/not-running/not-running.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      axios.get(`/server/${id}`).then((response) => {
+        return response.data;
+      }),
+    enabled: !!id,
+  });
+
+  useEffect(() => {
+    if (currentUser) {
+      const currentId = window.location.pathname
+        .replace(/\/$/, '')
```
Comment (jbouder) [inline ui/src/pages/not-running/not-running.tsx:27]: Trim any extra slashes from the end to ensure we get a proper split on strings

File: ui/src/utils/jupyterhub.ts [hunk 5]
```diff
@@ -93,6 +97,7 @@ export const getApps = (
         ...server,
         id: app.name,
         name: app.display_name,
+        url: server.url?.replace('/user/', '/hub/user/'),
```
Comment (jbouder) [inline ui/src/utils/jupyterhub.ts:100]: Fixes 404 error when clicking on stopped app

**Review by kildre — APPROVED** [review]
It looks good to me, tests passing and coverage good. No errors.

**Comment by aktech** [comment]
I am seeing this on a local nebari deployment, any ideas? @jbouder 

Screenshot 2024-07-15 at 3 06 30 pm

**Comment by jbouder** [comment]
> I am seeing this on a local nebari deployment, any ideas? @jbouder
> 
> Screenshot 2024-07-15 at 3 06 30 pm

Aside from this message, is the app loading as expected?

**Comment by aktech** [comment]
>Aside from this message, is the app loading as expected?

Yep, everything else seems to be running fine.

**Comment by jbouder** [comment]
> > Aside from this message, is the app loading as expected?
> 
> Yep, everything else seems to be running fine.

Cool. I think there is an overly specific check thats adding that message. I'll get an update pushed shortly.

**Comment by aktech** [comment]
It looks like the modal appears even when I just got off to clicking on the context menu, seems weird IMO, is that expected behaviour?

https://github.com/user-attachments/assets/5b5c32f7-1ccd-4f4c-8c92-b95265580176

**Comment by aktech** [comment]
We would need to handle the case when a shared app is stopped and the person with whom the app is shared, tried to access the app and then start it, in that case "starting the app" will fail. At the moment backend returns 500, but after this is merged https://github.com/nebari-dev/jhub-apps/pull/397 it will return 404. 

We're not allowing the shared user to start the app and in future we may, in that case it will return either 200 or 403.

So to summarise we need to handle the following cases:

- 200 (already handled)
- 404, 403: The error message should be something like you don't have permissions to start user "original-author"'s app, please ask the app author to start the app. (Same error for both cases since we're not allowing app starts by non-author for now)
- 500: Something went wrong, generic error.

The later two cases are not handled at the moment. I am happy to get this PR in and handle those in a follow up PR as this is a huge improvement anyway.

**Comment by jbouder** [comment]
> It looks like the modal appears even when I just got off to clicking on the context menu, seems weird IMO, is that expected behaviour?
> 
>  Screen.Recording.2024-07-16.at.9.43.49.am.mov

Good catch. Since we're adding a click event to the card, I need to make sure the events on the card and the context menu do not conflict. I fixed that in one place, but missed this one. Will get that fixed.

**Comment by jbouder** [comment]
> We would need to handle the case when a shared app is stopped and the person with whom the app is shared, tried to access the app and then start it, in that case "starting the app" will fail. At the moment backend returns 500, but after this is merged #397 it will return 404.
> 
> We're not allowing the shared user to start the app and in future we may, in that case it will return either 200 or 403.
> 
> So to summarise we need to handle the following cases:
> 
> * 200 (already handled)
> * 404, 403: The error message should be something like you don't have permissions to start user "original-author"'s app, please ask the app author to start the app. (Same error for both cases since we're not allowing app starts by non-author for now)
> * 500: Something went wrong, generic error.
> 
> The later two cases are not handled at the moment. I am happy to get this PR in and handle those in a follow up PR as this is a huge improvement anyway.

I think we can handle this separately as it may take a little more work. Can you create an issue with the specifics?

**Comment by jbouder** [comment]
> > It looks like the modal appears even when I just got off to clicking on the context menu, seems weird IMO, is that expected behaviour?
> > Screen.Recording.2024-07-16.at.9.43.49.am.mov
> 
> Good catch. Since we're adding a click event to the card, I need to make sure the events on the card and the context menu do not conflict. I fixed that in one place, but missed this one. Will get that fixed.

Fixed

**Review by aktech — APPROVED** [review]
Looks great to me. Verified the click context, works as expected.

**Pull request merged**